File: libs/evals/src/gap/evals/eval/metrics.py
"""Content quality scoring — compares base and GAP output artifacts."""

from __future__ import annotations

import difflib
import re
from collections import Counter
from pathlib import Path

from ..models import ContentQualityScore, ExperimentQuality

_GAP_MARKER_RE = re.compile(r"</?gap:target[^>]*>")


def strip_gap_markers(text: str) -> str:
    """Remove all <gap:target ...> and </gap:target> tags."""
    return _GAP_MARKER_RE.sub("", text)


def _token_f1(a: str, b: str) -> float:
    """Word-token F1 score between two texts."""
    ta = Counter(a.lower().split())
    tb = Counter(b.lower().split())
    common = sum((ta & tb).values())
    if common == 0:
        return 0.0
    precision = common / max(sum(tb.values()), 1)
    recall = common / max(sum(ta.values()), 1)
    return 2 * precision * recall / (precision + recall)


def _diff_line_counts(a: str, b: str) -> tuple[int, int]:
    """Count lines added and removed between two texts."""
    diff = list(difflib.unified_diff(a.splitlines(), b.splitlines(), lineterm=""))
    added = sum(1 for l in diff if l.startswith("+") and not l.startswith("+++"))
    removed = sum(1 for l in diff if l.startswith("-") and not l.startswith("---"))
    return added, removed


def _rouge_l(base: str, gap: str) -> float | None:
    """Compute ROUGE-L via ragas if available."""
    try:
        from ragas import SingleTurnSample
        from ragas.metrics import RougeScore

        scorer = RougeScore(rouge_type="rougeL")
        sample = SingleTurnSample(response=gap, reference=base)
        return scorer.single_turn_score(sample)
    except Exception:
        return None


def _bleu(base: str, gap: str) -> float | None:
    """Compute BLEU via ragas if available."""
    try:
        from ragas import SingleTurnSample
        from ragas.metrics import BleuScore

        scorer = BleuScore()
        sample = SingleTurnSample(response=gap, reference=base)
        return scorer.single_turn_score(sample)
    except Exception:
        return None


def score_turn(
    base_text: str,
    gap_text: str,
    turn: int,
    use_ragas: bool = False,
) -> ContentQualityScore:
    """Compute quality metrics for one turn pair."""
    gap_clean = strip_gap_markers(gap_text)
    seq_sim = difflib.SequenceMatcher(None, base_text, gap_clean).ratio()
    f1 = _token_f1(base_text, gap_clean)
    added, removed = _diff_line_counts(base_text, gap_clean)
    base_chars = len(base_text)
    gap_chars = len(gap_clean)
    delta_pct = ((gap_chars - base_chars) / base_chars * 100) if base_chars > 0 else 0.0

    rouge = _rouge_l(base_text, gap_clean) if use_ragas else None
    bleu = _bleu(base_text, gap_clean) if use_ragas else None

    return ContentQualityScore(
        turn=turn,
        sequence_similarity=round(seq_sim, 4),
        token_f1=round(f1, 4),
        base_char_count=base_chars,
        gap_char_count=gap_chars,
        char_delta_pct=round(delta_pct, 1),
        lines_added=added,
        lines_removed=removed,
        rouge_l=round(rouge, 4) if rouge is not None else None,
        bleu=round(bleu, 4) if bleu is not None else None,
    )


def score_experiment(
    base_output_dir: Path,
    gap_output_dir: Path,
    ext: str,
    use_ragas: bool = False,
) -> ExperimentQuality:
    """Score all turns for one experiment by reading artifacts from disk."""
    scores: list[ContentQualityScore] = []

    # Find matching turn files (turn-0, turn-1, ...)
    base_files = sorted(base_output_dir.glob(f"turn-*{ext}"))
    for bf in base_files:
        af = gap_output_dir / bf.name
        if not af.exists():
            continue
        turn_str = bf.stem.split("-")[1]
        turn = int(turn_str)
        scores.append(score_turn(bf.read_text(), af.read_text(), turn, use_ragas))

    if not scores:
        return ExperimentQuality(per_turn=[])

    mean_sim = sum(s.sequence_similarity for s in scores) / len(scores)
    mean_f1 = sum(s.token_f1 for s in scores) / len(scores)

    rouge_vals = [s.rouge_l for s in scores if s.rouge_l is not None]
    bleu_vals = [s.bleu for s in scores if s.bleu is not None]

    return ExperimentQuality(
        per_turn=scores,
        mean_sequence_similarity=round(mean_sim, 4),
        mean_token_f1=round(mean_f1, 4),
        mean_rouge_l=round(sum(rouge_vals) / len(rouge_vals), 4) if rouge_vals else None,
        mean_bleu=round(sum(bleu_vals) / len(bleu_vals), 4) if bleu_vals else None,
    )

File: apps/eval-cli/scorer/metrics.go
// Package scorer provides text quality scoring and LLM-as-judge evaluation.
package scorer

import (
	"math"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/urmzd/generative-artifact-protocol/eval-cli/gap"
)

// tokenF1 computes word-token F1 between two texts.
func tokenF1(a, b string) float64 {
	ta := wordCounts(strings.ToLower(a))
	tb := wordCounts(strings.ToLower(b))
	common := 0
	for k, va := range ta {
		if vb, ok := tb[k]; ok {
			common += min(va, vb)
		}
	}
	if common == 0 {
		return 0
	}
	sumA := sumValues(ta)
	sumB := sumValues(tb)
	precision := float64(common) / float64(max(sumB, 1))
	recall := float64(common) / float64(max(sumA, 1))
	return 2 * precision * recall / (precision + recall)
}

func wordCounts(text string) map[string]int {
	m := make(map[string]int)
	for _, w := range strings.Fields(text) {
		m[w]++
	}
	return m
}

func sumValues(m map[string]int) int {
	total := 0
	for _, v := range m {
		total += v
	}
	return total
}

// sequenceSimilarity computes character-level similarity ratio (like difflib.SequenceMatcher).
func sequenceSimilarity(a, b string) float64 {
	if len(a) == 0 && len(b) == 0 {
		return 1.0
	}
	lcsLen := lcs(a, b)
	return 2.0 * float64(lcsLen) / float64(len(a)+len(b))
}

// lcs computes the length of the longest common subsequence.
func lcs(a, b string) int {
	m, n := len(a), len(b)
	if m == 0 || n == 0 {
		return 0
	}
	// Space-optimized LCS.
	prev := make([]int, n+1)
	curr := make([]int, n+1)
	for i := 1; i <= m; i++ {
		for j := 1; j <= n; j++ {
			if a[i-1] == b[j-1] {
				curr[j] = prev[j-1] + 1
			} else {
				curr[j] = max(prev[j], curr[j-1])
			}
		}
		prev, curr = curr, prev
		clear(curr)
	}
	return prev[n]
}

// diffLineCounts counts lines added and removed between two texts.
func diffLineCounts(a, b string) (added, removed int) {
	linesA := strings.Split(a, "\n")
	linesB := strings.Split(b, "\n")

	setA := make(map[string]int)
	for _, l := range linesA {
		setA[l]++
	}
	setB := make(map[string]int)
	for _, l := range linesB {
		setB[l]++
	}

	for l, count := range setB {
		if countA, ok := setA[l]; ok {
			added += max(count-countA, 0)
		} else {
			added += count
		}
	}
	for l, count := range setA {
		if countB, ok := setB[l]; ok {
			removed += max(count-countB, 0)
		} else {
			removed += count
		}
	}
	return
}

// rougeL computes ROUGE-L F1 between reference and hypothesis.
func rougeL(reference, hypothesis string) float64 {
	refWords := strings.Fields(reference)
	hypWords := strings.Fields(hypothesis)
	if len(refWords) == 0 || len(hypWords) == 0 {
		return 0
	}
	lcsLen := lcsWords(refWords, hypWords)
	precision := float64(lcsLen) / float64(len(hypWords))
	recall := float64(lcsLen) / float64(len(refWords))
	if precision+recall == 0 {
		return 0
	}
	return 2 * precision * recall / (precision + recall)
}

func lcsWords(a, b []string) int {
	m, n := len(a), len(b)
	prev := make([]int, n+1)
	curr := make([]int, n+1)
	for i := 1; i <= m; i++ {
		for j := 1; j <= n; j++ {
			if a[i-1] == b[j-1] {
				curr[j] = prev[j-1] + 1
			} else {
				curr[j] = max(prev[j], curr[j-1])
			}
		}
		prev, curr = curr, prev
		clear(curr)
	}
	return prev[n]
}

// ScoreTurn computes quality metrics for one turn pair.
func ScoreTurn(baseText, gapText string, turn int) gap.ContentQualityScore {
	gapClean := gap.StripGAPMarkers(gapText)
	seqSim := sequenceSimilarity(baseText, gapClean)
	f1 := tokenF1(baseText, gapClean)
	added, removed := diffLineCounts(baseText, gapClean)
	baseChars := len(baseText)
	gapChars := len(gapClean)
	deltaPct := 0.0
	if baseChars > 0 {
		deltaPct = float64(gapChars-baseChars) / float64(baseChars) * 100
	}
	rl := rougeL(baseText, gapClean)

	return gap.ContentQualityScore{
		Turn:               turn,
		SequenceSimilarity: round4(seqSim),
		TokenF1:            round4(f1),
		BaseCharCount:      baseChars,
		GAPCharCount:       gapChars,
		CharDeltaPct:       math.Round(deltaPct*10) / 10,
		LinesAdded:         added,
		LinesRemoved:       removed,
		RougeL:             ptr(round4(rl)),
	}
}

// ScoreExperiment scores all turns for one experiment by reading artifacts from disk.
func ScoreExperiment(baseOutputDir, gapOutputDir, ext string) gap.ExperimentQuality {
	pattern := filepath.Join(baseOutputDir, "turn-*"+ext)
	baseFiles, _ := filepath.Glob(pattern)
	sort.Strings(baseFiles)

	var scores []gap.ContentQualityScore
	for _, bf := range baseFiles {
		name := filepath.Base(bf)
		af := filepath.Join(gapOutputDir, name)
		if _, err := os.Stat(af); err != nil {
			continue
		}
		baseData, _ := os.ReadFile(bf)
		gapData, _ := os.ReadFile(af)
		turn := parseTurnFromFilename(name)
		scores = append(scores, ScoreTurn(string(baseData), string(gapData), turn))
	}

	if len(scores) == 0 {
		return gap.ExperimentQuality{PerTurn: []gap.ContentQualityScore{}}
	}

	var sumSim, sumF1, sumRL float64
	rlCount := 0
	for _, s := range scores {
		sumSim += s.SequenceSimilarity
		sumF1 += s.TokenF1
		if s.RougeL != nil {
			sumRL += *s.RougeL
			rlCount++
		}
	}
	n := float64(len(scores))
	quality := gap.ExperimentQuality{
		PerTurn:                scores,
		MeanSequenceSimilarity: round4(sumSim / n),
		MeanTokenF1:            round4(sumF1 / n),
	}
	if rlCount > 0 {
		rl := round4(sumRL / float64(rlCount))
		quality.MeanRougeL = &rl
	}
	return quality
}

func parseTurnFromFilename(name string) int {
	// name like "turn-0.html"
	base := strings.TrimSuffix(name, filepath.Ext(name))
	parts := strings.SplitN(base, "-", 2)
	if len(parts) < 2 {
		return 0
	}
	n := 0
	for _, c := range parts[1] {
		if c >= '0' && c <= '9' {
			n = n*10 + int(c-'0')
		}
	}
	return n
}

func round4(v float64) float64 {
	return math.Round(v*10000) / 10000
}

func ptr[T any](v T) *T {
	return &v
}
